File: src/merge.cpp
/* Genotype Representation Graph Library (GRGL)
 * Copyright (C) 2024 April Wei
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
#include <args.hxx>
#include <fstream>
#include <iostream>
#include <string>

#include "grg_helpers.h"
#include "grgl/grg.h"
#include "grgl/serialize.h"
#include "grgl/visitor.h"
#include "util.h"

int main(int argc, char** argv) {
    args::ArgumentParser parser("Merge two GRGs.");
    args::HelpFlag help(parser, "help", "Display this help menu", {'h', "help"});
    args::Flag showStats(parser, "show-stats", "Show statistics about the GRG", {'s', "show-stats"});
    args::Flag noSimplify(parser, "no-simplify", "Simplify the resulting GRG", {'l', "no-simplify"});
    args::Flag noCombine(parser, "no-combine", "Do not combine nodes with same samplesets", {'c', "no-combine"});
    args::Positional<std::string> outfile(parser, "outfile", "The output file (resulting .grg)");
    args::PositionalList<std::string> inputs(parser, "inputs", "The input files (must be .grg)");
    try {
        parser.ParseCLI(argc, argv);
    } catch (args::Help&) {
        std::cout << parser;
        return 0;
    } catch (args::ParseError& e) {
        std::cerr << e.what() << std::endl;
        std::cerr << parser;
        return 1;
    } catch (args::ValidationError& e) {
        std::cerr << e.what() << std::endl;
        std::cerr << parser;
        return 1;
    }
    if (!inputs || !outfile || inputs->empty()) {
        std::cout << parser;
        return 0;
    }

    std::string firstFile;
    std::list<std::string> otherFiles;
    for (const auto& filename : *inputs) {
        if (firstFile.empty()) {
            firstFile = filename;
        } else {
            otherFiles.push_back(filename);
        }
    }
    grgl::MutableGRGPtr grg1 = grgl::loadMutableGRG(firstFile);
    if (!grg1) {
        return 2;
    }

    grg1->merge(otherFiles, !noCombine);

    if (showStats) {
        dumpStats(grg1);
    }

    std::ofstream outStream(*outfile, std::ios::binary);
    auto counts = grgl::writeGrg(grg1, outStream, !noSimplify);
    std::cout << "Wrote simplified GRG with:" << std::endl;
    std::cout << "  Nodes: " << counts.first << std::endl;
    std::cout << "  Edges: " << counts.second << std::endl;
    std::cout << "Wrote GRG to " << *outfile << std::endl;
    return 0;
}

File: pygrgl/clicmd/merge.py
# Genotype Representation Graph Library (GRGL)
# Copyright (C) 2025 April Wei
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# with this program.  If not, see <https://www.gnu.org/licenses/>.
from .common import which

from pygrgl import load_mutable_grg, save_grg

GRG_MERGE = which("grg-merge", required=True)


def add_options(subparser):
    subparser.add_argument("out_file", help="The output GRG file to create")
    subparser.add_argument("grg_file", nargs="*", help="The input GRG files")
    subparser.add_argument(
        "-C",
        "--no-combine",
        action="store_true",
        help="Don't combine nodes, keep all the nodes/edges from all graphs.",
    )
    subparser.add_argument(
        "-s",
        "--spec-file",
        help="A specification file listing all the GRGs to be merged. Each line of the file contains"
        "a single filename. Each filename can be suffixed with ':<position_offset>` to add the value "
        "<position_offset> to every mutation position in the GRG prior to merging.",
    )
    subparser.add_argument(
        "--use-samples",
        action="store_true",
        help="Use node equivalence via samples beneath each node, instead of just children. "
        "This reuses less hierarchy in the graph, but can produce a smaller graph. "
        "Slower and more RAM intensive.",
    )
    subparser.add_argument(
        "-v",
        "--verbose",
        action="store_true",
        help="Verbose output.",
    )


def merge_command(args):
    if len(args.grg_file) == 0 and args.spec_file is None:
        raise RuntimeError(
            "You must provide either a list of input GRGs on the command line,"
            " or --spec-file with a file containing a list of GRGs"
        )
    input_files = []
    adjust_positions = []
    if args.spec_file is not None:
        with open(args.spec_file, "r") as f:
            spec = list(filter(lambda s: len(s) > 0, map(str.strip, f)))
        for s in spec:
            s = s.split(":")
            input_files.append(s[0])
            if len(s) > 1:
                adjust_positions.append(int(s[1]))
        if adjust_positions and (len(input_files) != len(adjust_positions)):
            raise RuntimeError(
                "Must provide adjustment positions for all files, if any are provided"
            )
        if adjust_positions and adjust_positions[0] != 0:
            raise RuntimeError(
                "The first to-be-merged graph must have an adjustment position of 0"
            )
    else:
        input_files = args.grg_file
    if len(input_files) <= 1:
        raise RuntimeError("Cannot merge fewer than two graphs together.")
    target = load_mutable_grg(input_files[0])
    target.merge(
        input_files[1:],
        combine_nodes=not args.no_combine,
        use_sample_sets=True if args.use_samples else False,
        verbose=args.verbose,
        position_adjust=adjust_positions[1:],
    )
    save_grg(target, args.out_file)
